### PR TITLE
Adds CustomStateSet

### PR DESCRIPTION
**Pull request by rachelandrew**
Adds the `CustomStateSet` interface in Markdown. As this is an addition to `ElementInternals` this also adds the property page and link to that Interface in HTML.

Reviewer @jpmedley

### CODE REVIEW

File: files/en-us/web/api/customstateset/index.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`CustomStateSet`** interface of the {{domxref('Document_Object_Model','','',' ')}} stores a list of possible states for a custom element to be in, and allows states to be added and removed from the set.
+
+## Description
+
+An HTML form element, such as a checkbox has different _states_, "checked" and "unchecked". Likewise, developers creating custom elements need to assign possible states to these elements. The `CustomStateList` allows these states to be stored, and accessed from the custom element.
+
+A instance of `CustomStateList` is returned by {{domxref("ElementInternals.states")}}. The `CustomStateList` object is a {{jsxref("Set")}}, and therefore the properties and methods available on a `Set` can be used when working with a `CustomSetList`.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/index.md:19]: I do not believe this is literally an instance of Set. It's `add()` method [pushes to a variable called `list_`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/custom/custom_state_set.cc;l=80). I suspect that `list_` is an instance of the [list structure in the C++ Standard Template Library](https://www.geeksforgeeks.org/list-cpp-stl/). I can't the variable's initialization.

Comment (jpmedley) [reply files/en-us/web/api/customstateset/index.md:19]: [I found it](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/custom/custom_state_set.h;l=53). It's a [vector](https://www.geeksforgeeks.org/vector-in-cpp-stl/). Can you please explain this without reference to `Set`?

Comment (jpmedley) [reply files/en-us/web/api/customstateset/index.md:19]: I know the spec language makes it sound like this is a Set. I think it merely means that the same algorithms should be used.

File: files/en-us/web/api/customstateset/index.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+Each value stored in a `CustomStateList` is a `<dashed-ident>`, in the format `--mystate`.
+
+### Interaction with CSS
+
+States are stored as a `<dashed-ident>` as this format can then be accessed from CSS using the _custom state pseudo-class_.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/index.md:25]: Is that [this](https://developer.mozilla.org/en-US/docs/Web/CSS/:state)? Maybe add a link.

**Review by jpmedley — CHANGES_REQUESTED** [review]
Please see the comments I made on line 19 of the interface page. I don't believe your assessment of what this interface is is correct. I'd like to resolve this before I finish the review.

File: files/en-us/web/api/customstateset/index.md (hunk 2 above)
Comment (rachelandrew) [reply files/en-us/web/api/customstateset/index.md:25]: I think that page is incorrect, or maybe an old version looked like that. I'll check tomorrow and remove it if I'm right. The current spec has the names stored in the set used directly which is why they are a dashed-ident.

Comment (rachelandrew) [reply files/en-us/web/api/customstateset/index.md:25]: I spotted the discussion, I'll remove that page you found as it's incorrect https://github.com/WICG/custom-state-pseudo-class/issues/6

File: files/en-us/web/api/customstateset/index.md (hunk 1 above)
Comment (rachelandrew) [reply files/en-us/web/api/customstateset/index.md:19]: It describes it as `setlike`, I have one in another interface described as `maplike`. Sets and maps they may not be but in BCD we also have a whole bunch of methods and so on which can be used due to their set- or maplike-ness. 

So I don't know how to deal with this, other than to refer to set or map? Should we then create pages for each of these methods? If we don't, and also don't refer to set, how is someone supposed to know that's how they work with this?

You can see the BCD table in the preview: https://pr8093.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CustomStateSet

Comment (jpmedley) [reply files/en-us/web/api/customstateset/index.md:19]: MDN Helper should have built these pages for you and you would have done very little to them before submitting. Let me look into this.

Comment (rachelandrew) [reply files/en-us/web/api/customstateset/index.md:19]: @jpmedley did you get anywhere with this? I've seen other instances of these setlike and maplike interfaces so would be good to know how to deal with them, as the specs seem to assume they then just gain set and map functionality.

Comment (jpmedley) [reply files/en-us/web/api/customstateset/index.md:19]: Things got in my way. I'll look at it today.

File: files/en-us/web/api/customstateset/clear/index.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  - API
+  - Method
+  - Reference
+  - clear
+  - CustomStateSet
+browser-compat: api.CustomStateSet.clear
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`clear`** method of the {{domxref("CustomStateSet")}} interface removes all elements from the `CustomStateSet` object.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/clear/index.md:14]:
```suggestion
The **`clear()`** method of the {{domxref("CustomStateSet")}} interface removes all elements from the `CustomStateSet` object.
```

File: files/en-us/web/api/customstateset/delete/index.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  - API
+  - Method
+  - Reference
+  - delete
+  - CustomStateSet
+browser-compat: api.CustomStateSet.delete
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`delete`** method of the {{domxref("CustomStateSet")}} interface deletes a single value from the `CustomStateSet`.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/delete/index.md:14]:
```suggestion
The **`delete()`** method of the {{domxref("CustomStateSet")}} interface deletes a single value from the `CustomStateSet`.
```

File: files/en-us/web/api/customstateset/entries/index.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  - API
+  - Method
+  - Reference
+  - entries
+  - CustomStateSet
+browser-compat: api.CustomStateSet.entries
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`entries`** method of the {{domxref("CustomStateSet")}} interface returns a new {{jsxref("Iterator")}} object, containing an array of `[value,value]` for each element in the `CustomStateSet`.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/entries/index.md:14]: `[value,value]`? I don't understand.

File: files/en-us/web/api/customstateset/foreach/index.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  - API
+  - Method
+  - Reference
+  - forEach
+  - CustomStateSet
+browser-compat: api.CustomStateSet.forEach
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`forEach`** method of the {{domxref("CustomStateSet")}} interface executes a provided function for each value in the `CustomStateSet` object.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/foreach/index.md:14]:
```suggestion
The **`forEach()`** method of the {{domxref("CustomStateSet")}} interface executes a provided function for each value in the `CustomStateSet` object.
```

File: files/en-us/web/api/customstateset/has/index.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  - API
+  - Method
+  - Reference
+  - has
+  - CustomStateSet
+browser-compat: api.CustomStateSet.has
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`has`** method of the {{domxref("CustomStateSet")}} interface returns a {{jsxref("Boolean")}} asserting whether an element is present with the given value.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/has/index.md:14]:
```suggestion
The **`has()`** method of the {{domxref("CustomStateSet")}} interface returns a {{jsxref("Boolean")}} asserting whether an element is present with the given value.
```

File: files/en-us/web/api/customstateset/keys/index.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  - API
+  - Method
+  - Reference
+  - keys
+  - CustomStateSet
+browser-compat: api.CustomStateSet.keys
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`keys`** method of the {{domxref("CustomStateSet")}} interface is an alias for {{domxref("CustomStateSet.values")}}.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/keys/index.md:14]:
```suggestion
The **`keys()`** method of the {{domxref("CustomStateSet")}} interface is an alias for {{domxref("CustomStateSet.values")}}.
```

File: files/en-us/web/api/customstateset/values/index.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  - API
+  - Method
+  - Reference
+  - values
+  - CustomStateSet
+browser-compat: api.CustomStateSet.values
+---
+{{DefaultAPISidebar("DOM")}}
+
+The **`values`** method of the {{domxref("CustomStateSet")}} interface returns a new iterator object that yields the values for each element in the `CustomStateSet` object in insertion order.
```
Comment (jpmedley) [inline files/en-us/web/api/customstateset/values/index.md:14]:
```suggestion
The **`values()`** method of the {{domxref("CustomStateSet")}} interface returns a new iterator object that yields the values for each element in the `CustomStateSet` object in insertion order.
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: files/en-us/web/api/customstateset/entries/index.md (hunk 5 above)
Comment (rachelandrew) [reply files/en-us/web/api/customstateset/entries/index.md:14]: Exactly what it says really, it's a set so there isn't a key, just the same as regular set entries https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/entries

**Review by jpmedley — APPROVED** [review]

**Pull request merged**